src/ser.rs

```diff
@@ -429,14 +429,50 @@ pub struct CollectMap<'a, W> {
     ser: &'a mut Serializer<W>,
 }
 
-impl<'a, W> CollectMap<'a, W> {
+impl<'a, W> CollectMap<'a, W>
+where
+    W: enc::Write,
+{
     fn new(ser: &'a mut Serializer<W>) -> Self {
         Self {
             buffer: BufWriter::new(Vec::new()),
             entries: Vec::new(),
             ser,
         }
     }
+
+    fn serialize<T: Serialize + ?Sized>(
+        &mut self,
+        maybe_key: Option<&'static str>,
+        value: &T,
+    ) -> Result<(), EncodeError<W::Error>> {
+        // Instantiate a new serializer, so that the buffer can be re-used.
+        let mut mem_serializer = Serializer::new(&mut self.buffer);
+        if let Some(key) = maybe_key {
+            key.serialize(&mut mem_serializer)
+                .map_err(|_| EncodeError::Msg("Struct key cannot be serialized.".to_string()))?;
+        }
+        value
+            .serialize(&mut mem_serializer)
+            .map_err(|_| EncodeError::Msg("Struct value cannot be serialized.".to_string()))?;
+
+        self.entries.push(self.buffer.buffer().to_vec());
+        self.buffer.clear();
+
+        Ok(())
+    }
+
+    fn end(mut self) -> Result<(), EncodeError<W::Error>> {
+        // This sorting step makes sure we have the expected order of the keys. Byte-wise
+        // comparison gives us the right order as keys in DAG-CBOR are always (text) strings, hence
+        // have the same CBOR major type 3. The length of the string is encoded in the following
+        // bits. This means that a shorter string sorts before a longer string.
```

Comment:
```suggestion
        // This sorting step makes sure we have the expected order of the keys. Byte-wise
        // comparison over the encoded forms gives us the right order as keys in DAG-CBOR are
        // always (text) strings, hence have the same CBOR major type 3. The length of the string
        // is encoded in the prefix bits along with the major type. This means that a shorter string
        // always sorts before a longer string even with the compact length representation.
```

maybe this? "encoded forms" was the bit I overlooked so I put that in, but also tried to hint toward the mechanism in the second half of the comment